src/sympc/tensor/replicatedshare_tensor.py

```diff
@@ -605,32 +642,13 @@ def distribute_shares(
             return ValueError(
                 "Number of shares to be distributed should be same as number of parties"
             )
+        args = [
+            [shares, party_rank, session] for party_rank in range(session.nr_parties)
+        ]
 
-        parties = session.parties
-        nshares = len(parties) - 1
-
-        ptr_list = []
-        for i in range(len(parties)):
-            party_shares = []
-
-            for j in range(i, i + nshares):
-                share = shares[j % (nshares + 1)]
-
-                if isinstance(share, torch.Tensor):
-                    party_shares.append(share)
-
-                elif isinstance(share, ShareTensor):
-                    tensor = share.tensor
-                    party_shares.append(tensor)
-
-            tensor = ReplicatedSharedTensor(
-                party_shares,
-                config=Config(encoder_base=1, encoder_precision=0),
-                session_uuid=session.rank_to_uuid[i],
-            ).send(parties[i])
-            ptr_list.append(tensor)
-
-        return ptr_list
+        return parallel_execution(ReplicatedSharedTensor.distribute_shares_to_party)(
```

Comment:
have you checked that parallel execution is useful here? It creates sub processes that can be slow to launch, so sometimes non parellel (ie classic) execution is faster. I would suggest running X experiments and comparing before implementing this feature

Reply:
Oh alright. Will have a look.

Reply:
Yeah, classic was slightly better, I will switch back to classic.

Reply:
Yeah we can switch to the older one - I do consider that we might revisit this and do it like you have here.
For the moment we always create a Thread/Process pool for parallelization - I think that if we use 3-4 parties (with VirtualMachines -- the testing setup) it would not give better results because we spend more time creating the thread pool than doing the effective work.

What might change, if we switch to real multi-party over the network, it might be that this approach might give better results depending on the number of parties.

Reply:
I will be keeping the refactoring part. That is having a separate function to distribute shares to a party and an overall function to distribute it to all parties. In the future would just add a parallel_execution statement around it.